### PR TITLE
Add headscale module

### PR DESCRIPTION
**Pull request by kradalby**
###### Motivation for this change

This PR adds a NixOS module to configure [`headscale`](https://github.com/juanfont/headscale). headscale is already in the package repository.

This is my first PR, and my first module, so I will appreciate feedback, specially if something can be done in a better way.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/networking/headscale.nix
Comment (pennae) [inline nixos/modules/services/networking/headscale.nix:73]: since recently there is a `mkPackageOption` similar to `mkEnableOption` to make these options nicer to declare, if you like that :)

File: nixos/modules/services/networking/headscale.nix
Comment (pennae) [inline nixos/modules/services/networking/headscale.nix:306]: did the contents of this description go missing?

File: nixos/modules/services/networking/headscale.nix
Comment (pennae) [inline nixos/modules/services/networking/headscale.nix:362]:
```suggestion
              verification endpoint, and it will be listening on:
              :http = port 80
```
this will render the final line as normal text, if you want to make clear that this line is part of a config you should wrap it in the appropriate docbook tags (eg `<literal>`)

File: nixos/modules/services/networking/headscale.nix
Comment (pennae) [inline nixos/modules/services/networking/headscale.nix:402]:
```suggestion
    users.groups.headscale = mkIf (cfg.group == "headscale") { };
```
`mkIf` does not define the attribute it is assigned if the condition is false, so this can be simplified. same for other instances of this pattern.

File: nixos/modules/services/networking/headscale.nix
Comment (pennae) [inline nixos/modules/services/networking/headscale.nix:411]:
```suggestion
        isNormalUser = false;
```
service users are system users, not normal users.

File: nixos/modules/services/networking/headscale.nix
Comment (pennae) [inline nixos/modules/services/networking/headscale.nix:410]: this is the default for normal users, system users ideally shouldn't have a shell set for them at all.

File: nixos/modules/services/networking/headscale.nix
Comment (pennae) [inline nixos/modules/services/networking/headscale.nix:417]:
```suggestion
      after = [ "network-online.target" ];
```
`network.target` is not strong enough to ensure that connections to the outside world can be made

File: nixos/modules/services/networking/headscale.nix
Comment (pennae) [inline nixos/modules/services/networking/headscale.nix:422]:
```suggestion
          HEADSCALE_DB_PASS="$(head -n1 ${escapeShellArg cfg.database.passwordFile})"
```
common indentation in `''`-delimited strings is deleted, so it pays to indent according to structure :)

this (and the next variable) aren't exported. is that intentional?

File: nixos/modules/services/networking/headscale.nix
Comment (pennae) [inline nixos/modules/services/networking/headscale.nix:426]:
```suggestion
        HEADSCALE_OIDC_CLIENT_SECRET="$(head -n1 ${escapeShellArg cfg.openIdConnect.clientSecretFile})"
        exec ${cfg.package}/bin/headscale serve
```
need the `exec` to not keep the shell that runs this script around forever

File: nixos/modules/services/networking/headscale.nix
Comment (pennae) [inline nixos/modules/services/networking/headscale.nix:432]: `configFile` is a store path, so it already contains a hash. if there's a way to specify an alternate config file in the start script it wold be better to use that mechanism.

File: nixos/modules/services/networking/headscale.nix
Comment (pennae) [inline nixos/modules/services/networking/headscale.nix:452]: does this work as intended? systemd expects a relative name here. (it might even be a good idea to remove the `runDir` and `dataDir` options in favor of using `"headscale"` for both unconditionally)

File: nixos/modules/services/networking/headscale.nix
Comment (pennae) [inline nixos/modules/services/networking/headscale.nix:141]: is this a mandatory option? if so it's better to not give a default and get an error out of the module system if it's not set

File: nixos/modules/services/networking/headscale.nix
Comment (pennae) [inline nixos/modules/services/networking/headscale.nix:407]: this won't be needed if systemd creates the directory anyway.

File: nixos/modules/services/networking/headscale.nix
Comment (pennae) [inline nixos/modules/services/networking/headscale.nix:53]: this key (and all others that use `mkNotNull`) will be set to empty lists if the corresponding `cfg` attribute is null. can headscale handle that? if not it's better to use `// optionalAttrs (...) { ... }` chains to add these optional attributes.

**Review by pennae — CHANGES_REQUESTED** [review]
hiya! that's a pretty good attempt for a first module :)

few more high-level things:
 - there's a bunch of periods missing in descriptions
 - descriptions are not markdown, they're docbook. you'll have to use docbook markup :(
 - headscale seems to integrate a letsencrypt client. could we reuse `security.acme` functionality for this, or does it *have* to run its own client? would be nice to reuse existing modules instead of duplicating that (security-relevant) effort.

File: nixos/modules/services/networking/headscale.nix
Comment (aanderse) [reply nixos/modules/services/networking/headscale.nix:432]: You're looking for `restartTriggers` - that is the convention for what you are doing here.

File: nixos/modules/services/networking/headscale.nix
Comment (aanderse) [reply nixos/modules/services/networking/headscale.nix:452]: A big 👍 for that. Maybe we can even run this service as `DynamicUser`? Or not 🤷‍♂️

File: nixos/modules/services/networking/headscale.nix
Comment (aanderse) [reply nixos/modules/services/networking/headscale.nix:407]: Agreed. `createHome` is a "bad" way to provision directories.

**Comment by aanderse** [comment]
Looking good 👍

On top of what @pennae said I thought you might consider creating a `settings` option. It would add some flexibility to the module and might allow you to cut down on the number of options you provide.

File: nixos/modules/services/networking/headscale.nix
Comment (pennae) [reply nixos/modules/services/networking/headscale.nix:432]: yes, that would work just as well. (we just think it's nicer to not have services use implicit paths in `/etc` if at all possible)

Comment (kradalby) [reply nixos/modules/services/networking/headscale.nix:432]: Very happy to take guidance regarding symlinking the file to `/etc`. I tried to clarify why its done in a comment and I have not seen an obvious way to work around it to make the `headscale` CLI "find" it when a user is running it.

Will look into the restartTriggers.

File: nixos/modules/services/networking/headscale.nix
Comment (kradalby) [reply nixos/modules/services/networking/headscale.nix:437]: I have tried to read up on `DynamicUser`, and it _might_ be possible, but for the same reason as my `/etc` config linking and access to the socket, I am not sure if I can?

File: nixos/modules/services/networking/headscale.nix
Comment (kradalby) [inline nixos/modules/services/networking/headscale.nix:131]: I am actually contemplating if there should be a:

```
default = "${dataDir}/private.key";
```

My reasoning is that if you dont have a key, headscale will create one. ATM there is no headscale cli/tool to create this key to put it in your secret management.

I have filed a ticket for such a CLI option to exist: https://github.com/juanfont/headscale/issues/286

**Comment by kradalby** [comment]
> On top of what @pennae said I thought you might consider creating a `settings` option. It would add some flexibility to the module and might allow you to cut down on the number of options you provide.

That would be sensible, specially if we add options quicker to the upstream than to the module, I have tried to look for a good example for that pattern.

Would the `zigbee2mqtt` module be a good example? https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/misc/zigbee2mqtt.nix#L61

**Comment by aanderse** [comment]
> Would the zigbee2mqtt module be a good example?

That is a great example. The entire configuration file can be specified via nix attributes - very convenient.

File: nixos/modules/services/networking/headscale.nix
Comment (aanderse) [reply nixos/modules/services/networking/headscale.nix]: Maybe I'll poke at `DynamicUser` if I have some free time.

**Comment by kradalby** [comment]
> That is a great example. The entire configuration file can be specified via nix attributes - very convenient.

I did a variant of that to allow arbitrary overrides, what do you think about that?

I am leaning towards keeping a lot of the Nix "options" as they help fixing some of the poor configuration design we did in the beginning of `headscale`.

File: nixos/modules/services/networking/headscale.nix
Comment (pennae) [inline nixos/modules/services/networking/headscale.nix:32]: maybe also note that this also the group the service will run as, regardless of whether the default user config is used or not.

File: nixos/modules/services/networking/headscale.nix
Comment (pennae) [reply nixos/modules/services/networking/headscale.nix:68]: sounds like a good idea if having the key in secret management is not crucial (eg because it has to stable over long periods).

File: nixos/modules/services/networking/headscale.nix
Comment (pennae) [inline nixos/modules/services/networking/headscale.nix:77]: this being xml those will not render as two paragraphs unless explicitly marked as such with `<para>` :(

(really hope the transition to markdown will happen soon enough, docbook is horrid)

File: nixos/modules/services/networking/headscale.nix
Comment (pennae) [inline nixos/modules/services/networking/headscale.nix:333]: all the leaf options defined here can conflict/merge with user-provided settings. it's best to treat everything in this set as defaults, eg by `mkDefault`ing every leaf attribute. another commonly used pattern would be to generate the config not from `cfg.extraSettings`, but from `recursiveUpdate (<this set>) cfg.extraSettings`. each has its pros and cons.

File: nixos/modules/services/networking/headscale.nix
Comment (pennae) [inline nixos/modules/services/networking/headscale.nix:404]:
```suggestion
    users.users.headscale = mkIf (cfg.user == "headscale") {
```

File: nixos/modules/services/networking/headscale.nix
Comment (pennae) [inline nixos/modules/services/networking/headscale.nix:420]:
```suggestion
          export HEADSCALE_DB_PASS="$(head -n1 ${escapeShellArg cfg.database.passwordFile})"
```

File: nixos/modules/services/networking/headscale.nix
Comment (pennae) [inline nixos/modules/services/networking/headscale.nix:440]: does the group need write permissions? if we're not mistaken those only apply to adding/removing stuff from the directory, which doesn't factor into connecting to a unix socket

File: nixos/modules/services/networking/headscale.nix
Comment (pennae) [inline nixos/modules/services/networking/headscale.nix:443]: is `0755` safe here? if the runtime directory is restricted to owner/group only then the state dir should probably be as well.

File: nixos/modules/services/networking/headscale.nix
Comment (kradalby) [reply nixos/modules/services/networking/headscale.nix]: Yes, it is needed, but it can be rotated "at will" if a user desire. So it will be automatically generated if it doesnt exist.

**Comment by kradalby** [comment]
I _think_ I have managed to address most of the issues, I am still unsure if it could run with `DynamicUser` tho, but I think the rest of permissions and users has been streamlined with the reviewers help.

File: nixos/modules/services/networking/headscale.nix
Comment (pennae) [inline nixos/modules/services/networking/headscale.nix:59]:
```suggestion
        example = 443;
```

File: nixos/modules/services/networking/headscale.nix
Comment (pennae) [inline nixos/modules/services/networking/headscale.nix:216]: the given example still uses markdown syntax. 
```suggestion
            <literal>hostname.namespace.base_domain</literal> (e.g. 
            <literal>myhost.mynamespace.example.com</literal>).
```

File: nixos/modules/services/networking/headscale.nix
Comment (pennae) [inline nixos/modules/services/networking/headscale.nix:407]:
```suggestion
```
did this survive the `optionalAttrs -> mkIf` change? in any case the systemd service below should be in the mkIf, not outside of it.

File: nixos/modules/services/networking/headscale.nix
Comment (pennae) [inline nixos/modules/services/networking/headscale.nix:437]: doesn't allow group access at all any more. did you mean?
```suggestion
        RuntimeDirectoryMode = "0750";
```

**Review by pennae — COMMENTED** [review]
looks pretty good now apart from a couple things that look like typos :) also has to be added to `nixos/modules/module-list.nix` to be importated automatically.

as for `DynamicUser`: since this module uses a group to allow access to the daemon process that probably isn't possible. with `DynamicUser` you can't set the service to use a pre-existing group, so any access except by root would be blocked.

File: nixos/modules/services/networking/headscale.nix
Comment (kradalby) [reply nixos/modules/services/networking/headscale.nix:437]: Yes make sense 🤦 , I think being ill has fogged my brain a bit, apologies.

**Comment by pennae** [comment]
erf, eval is failing because `mkPackageOption` wasn't added to the default lib exports. will have to make a PR to fix that. in the meantime, can you squash everything down to two commits (one for adding the maintainer entry, one for the rest)?

**Comment by kradalby** [comment]
> in the meantime, can you squash everything down to two commits (one for adding the maintainer entry, one for the rest)?

Done, should I merge master in when #157301 is in?

**Comment by pennae** [comment]
rebasing would be better for bisectability, with a merge we'd have a commit that's known to not eval in the history

**Comment by cole-h** [comment]
@ofborg eval

File: nixos/modules/services/networking/headscale.nix
Comment (aanderse) [inline nixos/modules/services/networking/headscale.nix:24]:
```suggestion
        description = ''
          User account under which headscale runs.
          <note><para>
          If left as the default value this user will automatically be created
          on system activation, otherwise you are responsible for
          ensuring the user exists before the headscale service starts.
          </para></note>
        '';
```

File: nixos/modules/services/networking/headscale.nix
Comment (aanderse) [inline nixos/modules/services/networking/headscale.nix:32]:
```suggestion
        description = ''
          Group under which headscale runs.
          <note><para>
          If left as the default value this group will automatically be created
          on system activation, otherwise you are responsible for
          ensuring the user exists before the headscale service starts.
          </para></note>
        '';
```

File: nixos/modules/services/networking/headscale.nix
Comment (aanderse) [inline nixos/modules/services/networking/headscale.nix:314]: By convention we call this `settings`, not `extraSettings`.

File: nixos/modules/services/networking/headscale.nix
Comment (aanderse) [inline nixos/modules/services/networking/headscale.nix:54]:
```suggestion
        type = types.port;
```

File: nixos/modules/services/networking/headscale.nix
Comment (aanderse) [inline nixos/modules/services/networking/headscale.nix:135]:
```suggestion
          type = types.nullOr types.port;
```

**Review by aanderse — APPROVED** [review]
This PR LGTM :+1:

I have a couple minor suggestions which I think would be nice to add before merging, if you're interested.

Thanks!

File: nixos/modules/services/networking/headscale.nix
Comment (pennae) [reply nixos/modules/services/networking/headscale.nix:24]: there was a discussion a while back about maybe adding a `mkUserOption`/`mkGroupOption` with such a text. feels like an increasingly good idea to standardize this :)

Comment (kradalby) [reply nixos/modules/services/networking/headscale.nix:24]: mm, that sounds sensible!

Comment (aanderse) [reply nixos/modules/services/networking/headscale.nix:24]: For sure! But in the meantime let's not make good the enemy of perfect.

File: nixos/modules/services/networking/headscale.nix
Comment (kradalby) [reply nixos/modules/services/networking/headscale.nix]: Sorted

**Comment by kradalby** [comment]
Thanks, sorted the suggestions @aanderse, I think I just need to wait for the merger of #157301 to get rid of the error.

File: nixos/modules/services/networking/headscale.nix
Comment (pennae) [reply nixos/modules/services/networking/headscale.nix]: for sure! just as a reminder in case anyone wants to pick that up :D

**Comment by pennae** [comment]
in the spirit of "perfect is the enemy of good", if that PR sits for too long just go with the old style of package option. no use holding this up for long if it's that small a thing.

**Review by pennae — APPROVED** [review]

**Pull request merged**
